### PR TITLE
HTML 5 canonicalization

### PR DESCRIPTION
**Pull request by adon-at-work**
- enhance ContextParser to take event-driven hooks (instead of
requiring OOP inheritance)
- apply input stream pre-processing as required by HTML 5 spec
- rewrite html being processed will be parsed consistently by all
supported browsers
- disable IE conditional comments
- bug fixed due to equal sign detection that judges the use of
encodeURI/encodeURIComponent filter

### CODE REVIEW

**Comment by yahoocla** [comment]
CLA is valid!

**Comment by adon-at-work** [comment]
@neraliu , please help review. you might want to start with https://github.com/yahoo/secure-handlebars/blob/apply-input-preprocessing-and-purify-html/tests/unit/run-strong-context-parser.js
@yukinying , when you have time, would be great if you can take a look at https://github.com/yahoo/secure-handlebars/blob/apply-input-preprocessing-and-purify-html/src/strong-context-parser.js. we may consider moving some of the improvements made there to the upstream repo at context-parser.

**Comment by yukinying** [comment]
The use of nextPotentialState assumes that the parser would always parse the whole input string. If the caller use the parser in a byte by byte manner, nextPotentialState may not be valid.

File: dist/secure-handlebars.js [hunk 1]
```diff
@@ -8950,7 +8950,7 @@ Authors: Nera Liu <neraliu@yahoo-inc.com>
 var debug = require('debug')('cph');
 
 /* import the required package */
-var CustomizedContextParser = require('./customized-context-parser.js'),
+var ContextParser = require('./strong-context-parser.js'),
```
Comment (neraliu) [inline dist/secure-handlebars.js]: should we use strict-context-parser?

**Comment by adon-at-work** [comment]
> The use of nextPotentialState assumes that the parser would always parse the whole input string. If the caller use the parser in a byte by byte manner, nextPotentialState may not be valid.

yes, the assumption and common use case is that the input per parsePartial() call takes more than 1 char (e.g., given `<a href="`, `{{url}}`, `">`, {{url}}, `</a>`, the first, third and fifth chunks will be passed to parsePartial()). And for those depending on nextPotentialState will have to judge when it's equal to -1 (i.e., input[i+1] is out of bound.)

> Perhaps what we should do is to read "previousState" when the second char is parsed

It's largely due to performance concern so that I chose the look-ahead approach as is. The fact is that rewriting can alter the results of walk(). Hence, if we choose the look-back approach, we will literally need 2 iterations, i.e., first complete rewriting the whole string, then walk() from i = 0 again.

> , and append a space or a non visible separator? 

sorry i don't get this.

File: dist/secure-handlebars.js (hunk 1 above)
Comment (adon-at-work) [inline dist/secure-handlebars.js]: i'd like to keep it first. that line will become
`var ContextParser = require('context-parser');`
if the changes in this PR are eventually adopted to the upstream.

File: dist/secure-handlebars.js [hunk 2]
```diff
@@ -9346,24 +9364,38 @@ ContextParserHandlebars.prototype.analyzeAst = function(ast, stateObj, charNo) {
...
     /* make lastStates[0] and lastStates[1] the same as the tree has one branch */
-    ast.left.length > 0 && ast.right.length === 0? r.lastStates[1] = r.lastStates[0] : '';
-    ast.left.length === 0 && ast.right.length > 0? r.lastStates[0] = r.lastStates[1] : '';
-    debug("analyzeAst:["+r.lastStates[0].state+"/"+r.lastStates[1].state+"]");
+    // ast.left.length > 0 && ast.right.length === 0? r.lastStates[1] = r.lastStates[0] : '';
+    // ast.left.length === 0 && ast.right.length > 0? r.lastStates[0] = r.lastStates[1] : '';
+    // debug("analyzeAst:["+r.lastStates[0].state+"/"+r.lastStates[1].state+"]");
 
     // if the two branches result in different state
-    if (r.lastStates[0].state !== r.lastStates[1].state) {
```
Comment (neraliu) [inline dist/secure-handlebars.js]: as discussed, we need to enhance the state comparison under the principle that for those internal states affecting the future decision of customized filters added should be in the deepCompareState function

Comment (neraliu) [inline dist/secure-handlebars.js]: we would do the above mentioned in next code push.

File: dist/secure-handlebars.js [hunk 3]
```diff
@@ -8987,6 +8987,7 @@ var filter = {
 */
 var reURIContextStartWhitespaces = /^(?:[\u0000-\u0020]|&#[xX]0*(?:1?[1-9a-fA-F]|10|20);?|&#0*(?:[1-9]|[1-2][0-9]|30|31|32);?|&Tab;|&NewLine;)*/;
 var uriAttributeNames = {'href':1, 'src':1, 'action':1, 'formaction':1, 'background':1, 'cite':1, 'longdesc':1, 'usemap':1, 'poster':1, 'xlink:href':1};
+var reEqualSign = /(?:=|&#0*61;?|&#[xX]0*3[dD];?|&equals;)/;
```
Comment (neraliu) [inline dist/secure-handlebars.js]: good catch, some the general principle is the {ascii} and {html entities} for comparison

File: dist/secure-handlebars.js [hunk 4]
```diff
@@ -10141,5 +10055,699 @@ module.exports.create = overrideHbsCreate;
...
+
+    // deep copy the config to this.config
+    this.config = {};
+    for (k in config) {
+        this.config[k] = config[k];
+    }
+}
+
+/* inherit contextParser.FastParser */
+StrongContextParser.prototype = Object.create(htmlParser.prototype);
```
Comment (neraliu) [inline dist/secure-handlebars.js]: same naming comment. StrongCP vs StrictCP

File: dist/secure-handlebars.js (hunk 2 above)
Comment (adon-at-work) [inline dist/secure-handlebars.js]: Just to clarify. This PR or next?

File: dist/secure-handlebars.js [hunk 5]
```diff
@@ -10141,5 +10055,699 @@ module.exports.create = overrideHbsCreate;
...
+        args = [].slice.call(arguments, 1);
+        for (i = 0; (listener = listeners[i]); i++) {
+            listener.apply(self, args);
+        }
+    }
+    return self;
+};
+
+/**
+ * @function StrongContextStrongContextParser#parsePartial
```
Comment (neraliu) [inline dist/secure-handlebars.js]: some typo

File: dist/secure-handlebars.js [hunk 6]
```diff
@@ -10141,5 +10055,699 @@ module.exports.create = overrideHbsCreate;
...
+        self.emit('preWalk', lastState, i, endsWithEOF);
+        self.walk(i, self.input, endsWithEOF);
+        self.emit('postWalk', lastState, self.state, i, endsWithEOF);
+    }
+
+    return (self.output = self.input.join(''));
+};
+
+
+StrongContextParser.prototype.walk = function(i, input, endsWithEOF) {
```
Comment (neraliu) [inline dist/secure-handlebars.js]: i have some concern about the maintenance of this function, i understand that u r having the similar logic with the original walk function with different function signature. we should have next discussion to absorb it back to original CP.

File: dist/secure-handlebars.js [hunk 7]
```diff
@@ -10141,5 +10055,699 @@ module.exports.create = overrideHbsCreate;
...
+
+function Canonicalize(state, i, endsWithEOF) {
+    
+    this.emit('preCanonicalize', state, i, endsWithEOF);
+
+    var reCanonicalizeNeeded = true,
+        input = this.input,
+        chr = input[i], nextChr = input[i+1],
+        potentialState = this._getNextState(state, i, endsWithEOF),
+        nextPotentialState = this._getNextState(potentialState, i + 1, endsWithEOF);
```
Comment (neraliu) [inline dist/secure-handlebars.js]: do u need to have length check on i + 1?

Comment (neraliu) [inline dist/secure-handlebars.js]: similar length check question below whenever we "i + X" in the code

Comment (adon-at-work) [inline dist/secure-handlebars.js]: nextChr = input[i + 1] will pick up undefined
_getNextState() have done i >= length check, if so, returns -1

the subsequent logic only read from these values for checks. would be fine

File: dist/secure-handlebars.js (hunk 6 above)
Comment (adon-at-work) [inline dist/secure-handlebars.js]: agree

**Pull request merged**